tasks/main.yml

```diff
@@ -0,0 +1,38 @@
+- name: set authorized keys
+  authorized_key:
+    user: "{{ passwordless_ssh_user }}"
+    state: present
+    key: "{{ lookup('file', passwordless_ssh_public_key) }}"
+  become: "{{ passwordless_ssh_become }}"
+  become_user: "{{ passwordless_ssh_become_user }}"
+
+- name: copy private key to remote SSH dir
+  copy:
+    src: "{{ passwordless_ssh_private_key }}"
+    dest: "{{ passwordless_ssh_user_home }}/.ssh/\
+           {{ passwordless_ssh_remote_key_file }}"
+    mode: 0600
+    owner: "{{ passwordless_ssh_user }}"
```

Comment:
Since this bit is security related it might be a good idea to be explicit about the **group** ownership, as well as the selinux file context, which I believe should be as follows:

`Context: unconfined_u:object_r:ssh_home_t:s0`

Reply:
Group is denied permission to read by `mode: 0600`, so I figured the system's default setgid behavior based on users should be OK there. And I don't have any idea what you just said about SELinux.

Reply:
With selinux enabled, files have extended attributes that can be seen by running stat: ~/.ssh/id_rsa, for example.  Most ansible modules (such as the copy module) that perform file operations support this stuff, usually with parameters prefixed with "se".  

I do wonder if this operation, without any selinux parameters, will inherit the correct file contexts, or if being explicit is the way to go here.  We could try testing it out and compare what ansible does vs a manually-generated SSH key on a selinux system using the "stat" test.  @seandst knows some selinux stuff, so I'm curious as to what his thoughts are here.

Reply:
An option for testing would be to touch the file with the `file` module and see if it just does the right thing with selinux contexts. Based on the `file` module [docs](https://docs.ansible.com/ansible/latest/modules/file_module.html), I believe that if no selinux settings are passed to `file` and if the system supports selinux, the module will use the system defaults for selinux stuff.

The **seuser** arg's doc says "User part of SELinux file context. *Will default to system policy, if applicable.* If set to _default, it will use the user portion of the policy if available." (emphasis mine). Both the **serole** and **seuser** args say that they are set to their respective "...part of SELinux file context, _default feature works as for seuser." It's not clear to me that both **serole** and **seuser** will default to the system policy like **seuser* does, but poking in the guts of `ansible.module_utils.basic.AnsibleModule` gives me a little more confidence that just poking the `known_hosts` file before invoking the `known_hosts` modules ought to be good enough (assuming the module works the way you need it to otherwise). Unfortunately, the `known_hosts` module does not invoke the file handling included in `AnsibleModule`, so I think both `file` and `known_hosts` would have to be used to ensure the file is created with the correct selinux context info.

(sudo) `restorecon ~/.ssh/known_hosts` should set the file's selinux contexts to what they should be for easier testing.

Reply:
I ran a manual test to check the behavior here, and used this file as a reference for proper selinux contexts:

`/etc/selinux/targeted/contexts/files/file_contexts.homedirs`

which gives:

`/home/[^/]+/\.ssh(/.*)? unconfined_u:object_r:ssh_home_t:s0`

I can confirm that the test user `~/.ssh` directory has the correct context set after this role has been used.  The `/root/.ssh` directory has a different default context of `system_u:object_r:ssh_home_t:s0`, which is also handled correctly by this role.